docs: understanding datamodel and REST API (#3913)
inveniosoftware/invenio: docs/tutorials/understanding-the-rest-api.rst

..
    This file is part of Invenio.
    Copyright (C) 2018 CERN.

    Invenio is free software; you can redistribute it and/or modify it
    under the terms of the MIT License; see LICENSE file for more details.

Understanding the REST API
==========================

> In the :doc:`previous section <./understanding-datamodels>` , the different parts involved in a datamodel were explained. In this section we will show the process followed inside the RESTful API when creating and reading a record.

Reviewer: is space needed between `<./understanding-datamodels>` and `,`?

.. code-block:: shell

    curl -k --header "Content-Type: application/json" \
        --request POST \
        --data '{"title":"Some title", "contributors": [{"name": "Doe, John"}]}' \
        https://localhost:5000/api/records/?prettyprint=1

We will now explain six steps, that correspond to the six big boxes of the previous diagram.

The first step is to load the record, which as its name implies, is performed by the **loader**. At the beginning of this section the `json_pid_checker` was configured as the `record_loader` for `application/json` requests, therefore, it is the one used in this case.

The record will be parsed by the `RecordMetadataSchemaJSONV1`. This schema only checks that the PID value is injected if needed (mostly for PUT operations). Nonetheless, you can create your own `Marshmallow` schema to fit your needs.

The next step is to check if the user has permissions to create records, by default all authenticated users are allowed. You can see more about access control in the `managing access section </docs/tutorial-module/managing-access/index>`_.

> Aftewards, the record is minted. This action again as its name implies, is carried out by the **minter**. It adds to the record the corresponding value to the attribute configured via `PIDSTORE_RECID_FIELD <https://github.com/inveniosoftware/invenio-pidstore/blob/master/invenio_pidstore/config.py#L11>`_. For example, increasing by one its value.

Reviewer 2: Generally avoid linking to GitHub repository code, since e.g. now the line numbers have changed. It's better to link to classes in docs or use `intersphinx`

Subsequently, the record has to be persisted to the SQL database. However, before it is inserted, it is **validated**. Meaning, that the data will be cross-checked against the **JSONSchema**.

> The next step, is to index it in Elasticsearch so the record can be searched. The record will be sent to the specified index (in the `$schema` field of the data) or the default if the attribute is not present. The validation of the data against the mapping is performed by Elasticsearch itself. In order to avoid errors and exceptions the JSONSchema should reflect the same data structure than the **ES mapping**.

Reviewer 2: ```suggestion
The next step, is to index it in Elasticsearch so the record can be searched. The record will be sent to the specified index (in the ``$schema`` field of the data) or the default index if the attribute is not present. The validation of the data against the mapping is performed by Elasticsearch itself. In order to avoid errors and exceptions, the JSONSchema should reflect the same data structure as the **ES mapping**.
```

> So far, the record has been loaded (deserialized), minted, persisted and indexed. Therefore, it is only missing the last step: `serialization`. This operation is carried out by the **serializer**. In this case, the `json_v1_response <https://github.com/inveniosoftware/invenio-records-rest/blob/master/invenio_records_rest/serializers/__init__.py#L20>`_ is configured as serializer for responses. It will apply the corresponding `Marshmallow` schema and send the created record in the HTTP reponse with code 201.

Reviewer 2: ```suggestion
So far, the record has been loaded (deserialized), minted, persisted and indexed. Therefore, it is only missing the last step: ``serialization``. This operation is carried out by the **serializer**. In this case, the `json_v1_response <https://github.com/inveniosoftware/invenio-records-rest/blob/master/invenio_records_rest/serializers/__init__.py#L20>`_ is configured as the serializer for responses. It will apply the corresponding ``Marshmallow`` schema and return the created record as JSON in the HTTP response with code 201.
```

Reading a record
----------------

> In the following flowchart, the process that a record goes through upon creation is shown:

Reviewer: it is not creation, right?

.. image:: ../_static/diagrams/read_record.png
    :align: center
    :scale: 80%

> `Note: If it is not displayed correctly click on it to be able to zoom in.`

Reviewer 2: ```suggestion
`Note: If the image is not displayed correctly click on it to be able to zoom in.`
```

> The process to read a record is a bit more simple than that of cration explained above. In order to continue with the previous example, we would perform a GET operation:

Reviewer 2: ```suggestion
The process to read a record is a bit more simple than that of creation explained above. In order to continue with the previous example, we would perform a GET operation:
```

.. code-block:: shell

    curl -k --header "Content-Type: application/json" \
        https://localhost:5000/api/records/1?prettyprint=1

This will hit the `item_route (/records/<pid(recid):pid_value>) <https://github.com/inveniosoftware/invenio-records-rest/blob/master/invenio_records_rest/views.py#L692>`_ of the REST API.

> The next step is to obtain the record from the PID (`recid`) given in the URL. This is done by the **resolver**, which maps the given persistent identifier to an object in the database, and returns it.

Reviewer 2: ```suggestion
The next step is to obtain the record from the PID (``recid``) given in the URL. This is done by the **resolver**, which maps the given persistent identifier to an object in the database, and returns it.
```

Afterwards, it has to be checked if the user has permissions to read the requested record, by default all authenticated users are allowed. You can see more about access control in `managing access section </docs/tutorial-module/managing-access/index>`_.

> Finally the record will be **serialized**. As before, the `json_v1_response <https://github.com/inveniosoftware/invenio-records-rest/blob/master/invenio_records_rest/serializers/__init__.py#L20>`_ is configured as serializer for responses. It will apply the corresponding `Marshmallow` schema and send the created record in the HTTP reponse with code 200.

Reviewer 2: ```suggestion
Finally the record will be **serialized**. As before, the `json_v1_response <https://github.com/inveniosoftware/invenio-records-rest/blob/master/invenio_records_rest/serializers/__init__.py#L20>`_ is configured as the serializer for responses. It will apply the corresponding ``Marshmallow`` schema and send the created record as JSON in the HTTP response with code 200.
```
